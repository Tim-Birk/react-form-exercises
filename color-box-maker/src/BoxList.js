import React, { useState } from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import { v4 as uuid } from 'uuid';
import './BoxList.css';

const BoxList = () => {
  const [boxes, setBoxes] = useState([
    { id: uuid(), color: 'green', width: 100, height: 115 },
    { id: uuid(), color: 'yellow', width: 90, height: 85 },
    { id: uuid(), color: 'red', width: 190, height: 185 },
    { id: uuid(), color: 'black', width: 175, height: 140 },
  ]);

  const addBox = (color, width, height) => {
    const newBox = { id: uuid(), color, width, height };
    setBoxes((oldBoxes) => [...oldBoxes, newBox]);
  };

  const removeBox = (id) => {
    setBoxes(boxes.filter((box) => box.id !== id));
  };

  return (
    <>
      <h1>Box List</h1>
      <NewBoxForm addBox={addBox} />
      <div className='BoxList'>
        {boxes.map((box) => (
          <Box
            key={box.id}
            id={box.id}
            color={box.color}
            width={box.width}
            height={box.height}
            removeBox={removeBox}
          />
        ))}
      </div>
    </>
  );
};

export default BoxList;
