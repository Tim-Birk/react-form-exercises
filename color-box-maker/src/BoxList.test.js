import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

it('renders without crashing', function () {
  render(<BoxList />);
});

it('matches snapshot', function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it('can add a new box', function () {
  const { getByLabelText, queryByText, queryByTestId } = render(<BoxList />);

  // no box yet
  expect(queryByTestId('Box')).not.toBeInTheDocument();

  const colorInput = getByLabelText('Color:');
  const heightInput = getByLabelText('Height:');
  const widthInput = getByLabelText('Width:');
  const submitBtn = queryByText('Add New');

  // fill out the form
  fireEvent.change(colorInput, { target: { value: 'purple' } });
  fireEvent.change(heightInput, { target: { value: 100 } });
  fireEvent.change(widthInput, { target: { value: 120 } });
  fireEvent.click(submitBtn);

  // box exists!
  expect(queryByTestId('Box')).toBeInTheDocument();
  // box styled correctly
  expect(queryByTestId('Box')).toHaveStyle(`
      background-color: purple;
      height: 100px;
      width: 120px;
    `);
});

it('can remove a new box', function () {
  const { getByLabelText, queryByText, queryByTestId } = render(<BoxList />);

  // no box yet
  expect(queryByTestId('Box')).not.toBeInTheDocument();

  const colorInput = getByLabelText('Color:');
  const heightInput = getByLabelText('Height:');
  const widthInput = getByLabelText('Width:');
  const submitBtn = queryByText('Add New');

  // fill out the form
  fireEvent.change(colorInput, { target: { value: 'purple' } });
  fireEvent.change(heightInput, { target: { value: 100 } });
  fireEvent.change(widthInput, { target: { value: 120 } });
  fireEvent.click(submitBtn);

  // box exists
  expect(queryByTestId('Box')).toBeInTheDocument();
  const boxDeleteBtn = queryByTestId('Box-delete');
  // box delete button exists
  expect(boxDeleteBtn).toBeInTheDocument();
  fireEvent.click(boxDeleteBtn);
  // box removed
  expect(queryByTestId('Box')).not.toBeInTheDocument();
});

it('can add a new box', function () {
  const { getByLabelText, queryByText, queryByTestId } = render(<BoxList />);

  // no box yet
  expect(queryByTestId('Box')).not.toBeInTheDocument();

  const colorInput = getByLabelText('Color:');
  const heightInput = getByLabelText('Height:');
  const widthInput = getByLabelText('Width:');
  const submitBtn = queryByText('Add New');

  // fill out the form
  fireEvent.change(colorInput, { target: { value: 'purple' } });
  fireEvent.change(heightInput, { target: { value: 100 } });
  fireEvent.change(widthInput, { target: { value: 120 } });
  fireEvent.click(submitBtn);

  // box exists!
  expect(queryByTestId('Box')).toBeInTheDocument();
  // box styled correctly
  expect(queryByTestId('Box')).toHaveStyle(`
      background-color: purple;
      height: 100px;
      width: 120px;
    `);
});
