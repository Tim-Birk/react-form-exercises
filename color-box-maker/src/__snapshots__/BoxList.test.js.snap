// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h1>
    Box List
  </h1>
  <form>
    <div>
      <label
        for="color"
      >
        Color:
      </label>
      <input
        id="color"
        name="color"
        value=""
      />
    </div>
    <div>
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value=""
      />
      <span>
        px
      </span>
    </div>
    <div>
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value=""
      />
      <span>
        px
      </span>
    </div>
    <button>
      Add New
    </button>
  </form>
  <div
    class="BoxList"
  >
    <div
      class="Box"
    >
      <div
        data-testid="Box-green-100-115"
        style="background-color: green; width: 100px; height: 115px;"
      />
      <button
        data-testid="Box-delete-green-100-115"
      >
        X
      </button>
    </div>
    <div
      class="Box"
    >
      <div
        data-testid="Box-yellow-90-85"
        style="background-color: yellow; width: 90px; height: 85px;"
      />
      <button
        data-testid="Box-delete-yellow-90-85"
      >
        X
      </button>
    </div>
    <div
      class="Box"
    >
      <div
        data-testid="Box-red-190-185"
        style="background-color: red; width: 190px; height: 185px;"
      />
      <button
        data-testid="Box-delete-red-190-185"
      >
        X
      </button>
    </div>
    <div
      class="Box"
    >
      <div
        data-testid="Box-black-175-140"
        style="background-color: black; width: 175px; height: 140px;"
      />
      <button
        data-testid="Box-delete-black-175-140"
      >
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
